clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -443,7 +443,9 @@ boolean joinGroupIfNeeded(final Timer timer) {
                     stateSnapshot = this.state;
                 }
 
-                if (!generationSnapshot.equals(Generation.NO_GENERATION) && stateSnapshot == MemberState.STABLE) {
+                if ((generationSnapshot.generationId != Generation.NO_GENERATION.generationId ||
+                    !generationSnapshot.memberId.equals(Generation.NO_GENERATION.memberId)) &&
+                    stateSnapshot == MemberState.STABLE) {
```

Comment:
Before this change, we'll always reset generation object to `NO_GENERATION`, but now, we'll have some cases that only reset generation ID. Since after rebalance complete, consumer should have a valid generation ID. No generation ID (i.e. -1) also means the consumer needs to rejoin group. So change the if condition here.

Reply:
Is the `||` in this condition correct? I thought that we would consider the rebalance successful only if we have a valid generation and a valid member id. Am I missing something?

Reply:
You are right! Updated.